### PR TITLE
CNS-2791: Add a remote configurator that pulls changes from CB Console and applies them to the CR

### PR DESCRIPTION
**Pull request by ltsonov-cb**
This MR adds a secondary "loop" within the operator, in addition to the main one that is responsible to maintain the agent.

Structure:
* A new controller that is only responsible to run the "reconfiguration" code every X seconds/minutes. It will also do exponential backoff if that fails to avoid clusters swarming us with requests for pending changes.
* A "configurator" that handles the main logic - getting the CR, reading and pending changes, changing the CR, etc
* Two helpers that validate a change (i.e. is it even OK to do it? - this should also be done on the backend) and apply it. Makes testing easier
* The feature is still incomplete. The API for changes is "mocked" and returns random changes instead. This is so a developer can try it out locally. It will basically reconfigure the CR every N seconds right now with random (valid) values. 
* Due to the above bullet, the feature is currently _disabled_ by default (requires env var to run). But in the final version it will be _enabled_ by default and controlled by the CR. The latter is implemented so we only have to remove the env var before release. 

The major changes are under `remote_configuration` package so I'd recommend starting the review from there and going `Controller->Configurator->Change_applier`.

Other changes:
* ApiGateway and AccessToken were required in both places so I extracted those from the controller's code
* The cluster identifier is currently propagated to ApiGateway calls. But we should refactor this code a bit since other cluster ID values (group+name) are part of the apigateway itself and it's hard to follow. 

Separate MR will handle:
* Implementing the ConfigChanges API client
* More docs

### CODE REVIEW

**Review by BenRub — CHANGES_REQUESTED** [review]

File: remote_configuration/change_applier.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	resetVersion := func(ptrToField *string) {
+		if ptrToField != nil && *ptrToField != "" {
+			*ptrToField = ""
+		}
+	}
+
+	if change.AgentVersion != nil {
+		cr.Spec.Version = *change.AgentVersion
+
+		resetVersion(&cr.Spec.Components.Basic.Monitor.Image.Tag)
```
[the remark this answers is not in the document]
Comment (ltsonov-cb) [reply remote_configuration/change_applier.go]: Cool, didn't think of that. Implemented and pushed.

File: remote_configuration/change_applier_test.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	"github.com/stretchr/testify/assert"
+	cbcontainersv1 "github.com/vmware/cbcontainers-operator/api/v1"
+	"github.com/vmware/cbcontainers-operator/cbcontainers/models"
+	"github.com/vmware/cbcontainers-operator/remote_configuration"
+	"math/rand"
+	"strconv"
+	"testing"
+)
+
+func TestFeatureTogglesAreAppliedCorrectly(t *testing.T) {
```
Comment (skaneti1) [inline remote_configuration/change_applier_test.go:14]: I think this test tried to save code by wrapping the tests with `appliedChangeTest`, iterating them and run them, but since the actual test is only:
```go
remote_configuration.ApplyConfigChangeToCR(testCase.change, &testCase.initialCR)
testCase.assertFinalCR(t, testCase.initialCR)
```

I think this wrapping and "code saving" didn't save a lot of code, but did made the test more messy and harder to read and understand. 

I think simple 4 t.Run tests with the 4 test cases will be clearer and probably will resolve with close code lines

File: api/v1/cbcontainersagent_types.go [hunk 3]
```diff
@@ -126,6 +129,14 @@ type CBContainersProxySettings struct {
 	NoProxySuffix *string `json:"noProxySuffix,omitempty"`
 }
 
+// CBContainersRemoteConfigurationSettings holds settings for the operator/agent's feature to apply configuration changes via the Carbon black console
+type CBContainersRemoteConfigurationSettings struct {
+	// EnabledForAgent turns the feature to change agent configuration remotely (as opposed to operator configuration)
+	//
+	// +kubebuilder:default:=true
```
[the remark this answers is not in the document]
Comment (ltsonov-cb) [reply api/v1/cbcontainersagent_types.go:136]: It's disabled by default during development  only (via env var), then the env var will be removed and this value will be canonical

File: cbcontainers/communication/gateway/api_gateway.go [hunk 4]
```diff
@@ -165,3 +168,43 @@ func (gateway *ApiGateway) GetCompatibilityMatrixEntryFor(operatorVersion string
...
+	// TODO: Real implementation with CNS-2790
+	c := randomRemoteConfigChange()
+	if c != nil {
+		return []models.ConfigurationChange{*c}, nil
+
+	}
+	return nil, nil
+}
+
+func (gateway *ApiGateway) UpdateConfigurationChangeStatus(context.Context, models.ConfigurationChangeStatusUpdate) error {
```
[the remark this answers is not in the document]
Comment (ltsonov-cb) [reply cbcontainers/communication/gateway/api_gateway.go:206]: Done

**Review by skaneti1 — APPROVED** [review]
